### PR TITLE
eth/tracers: extend tracers for `CREATE2`

### PR DESCRIPTION
**Pull request by YaoZengzeng**
ref: #17196

### CODE REVIEW

**Comment by holiman** [comment]
:+1:

**Comment by holiman** [comment]
If we just merge this, it will have no effect, the `assets.go` needs to be regenerated aswell, using `go-bindata`. 
See https://github.com/ethereum/go-ethereum/blob/beb141aa9bdf3513e241bf3db802a0bc796ee0fa/eth/tracers/internal/tracers/assets.go

**Comment by karalabe** [comment]
You just need to do `go generate` in the tracers folder. Do `go get -u github.com/kevinburke/go-bindata/go-bindata` beforehand to make sure you have the latest bindata.

**Comment by YaoZengzeng** [comment]
@holiman @karalabe Updated

File: eth/tracers/internal/tracers/prestate_tracer.js [hunk 1]
```diff
@@ -85,7 +85,7 @@
 			case "EXTCODECOPY": case "EXTCODESIZE": case "BALANCE":
 				this.lookupAccount(toAddress(log.stack.peek(0).toString(16)), db);
 				break;
-			case "CREATE":
+			case "CREATE": case "CREATE2":
```
Comment (karalabe) [inline eth/tracers/internal/tracers/prestate_tracer.js]: This is not correct. The code below uses `toContract(from, db.getNonce(from))` to generate the contract address, but CREATE2 has a different contract address generation scheme. You'll need to implement a `toContract2` similar to https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/tracer.go#L354

**Comment by YaoZengzeng** [comment]
@karalabe Updated, actually I know nothing about js and I implement `toContract2` with the knowledge of existing functions like `toContract` and `toAddress` 😂 

PTAL if it is correct :)

File: eth/tracers/tracer.go [hunk 2]
```diff
@@ -365,6 +365,21 @@ func New(code string) (*Tracer, error) {
...
+		if ptr, size := ctx.GetBuffer(-3); ptr != nil {
+			from = common.BytesToAddress(makeSlice(ptr, size))
+		} else {
+			from = common.HexToAddress(ctx.GetString(-3))
+		}
+		salt := common.HexToHash(ctx.GetString(-2))
+		code := common.CopyBytes(makeSlice(ctx.GetBuffer(-1)))
+		ctx.Pop3()
+
+		contract := crypto.CreateAddress2(from, salt, code)
```
Comment (holiman) [inline eth/tracers/tracer.go]: This is now wrong, `CreateAddress2` now uses `codeHash` , not `code`.

**Comment by holiman** [comment]
@YaoZengzeng could you update it to use codehash instead?

**Comment by YaoZengzeng** [comment]
I almost unfamiliar with the code base now.

I will close this PR and maybe @holiman could make another PR to fix it.

Sorry for that.

**Pull request closed without merging**